arangod/GeoIndex/Covering.h

```diff
@@ -0,0 +1,124 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <s2/s2cap.h>
+#include <s2/s2cell_id.h>
+#include <s2/s2region.h>
+#include <s2/s2region_coverer.h>
+
+#include <queue>
+#include <vector>
+
+#include "Geo/GeoParams.h"
+#include "Geo/Utils.h"
+#include "VocBase/Identifiers/LocalDocumentId.h"
+
+namespace arangodb {
+namespace geo_index {
+
+/// @brief Helper class to build a simple covering query iterator.
+/// Will return findings not sorted by anything, will filter according
+/// to the filterType. Should be storage engine agnostic.
+class CoveringUtils {
+ public:
+  /// @brief Type of documents buffer
+  typedef std::deque<LocalDocumentId> GeoDocumentsQueue;
+
+  explicit CoveringUtils(geo::QueryParams&& params) noexcept;
+  ~CoveringUtils();
+
+ public:
+  inline geo::FilterType filterType() const { return _params.filterType; }
+
+  inline geo::ShapeContainer const& filterShape() const {
+    return _params.filterShape;
+  }
+
+  /// @brief all intervals are covered, no more buffered results
+  bool isDone() const { return _buffer.empty() && _allIntervalsCovered; }
+
+  /// @brief has buffered results
+  inline bool hasNext() const { return !_buffer.empty(); }
+
+  inline size_t bufferSize() const { return _buffer.size(); }
+
+  LocalDocumentId const& getNext() const {
+    TRI_ASSERT(hasNext());
+    return _buffer.front();
+  }
+
+  inline void next() {
+    TRI_ASSERT(hasNext());
+    _buffer.pop_front();
+  }
```

Comment:
Is there a reason why we use `pop_front` instead of maintaining an index? From what I can tell, it looks like that `_buffer` is filled once with everything that is found in one interval and then consumed completely. One could even use a vector here instead of a deque, right?